lib/srv/db/sqlserver/engine.go

```diff
@@ -99,15 +101,78 @@ func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Sessio
 		return trace.Wrap(err)
 	}
 
-	// Start proxying packets between client and server.
-	err = e.proxy(ctx, serverConn)
-	if err != nil {
-		return trace.Wrap(err)
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	defer e.Audit.OnSessionEnd(e.Context, sessionCtx)
+
+	clientErrCh := make(chan error, 1)
+	serverErrCh := make(chan error, 1)
+	go e.receiveFromClient(e.clientConn, serverConn, clientErrCh, sessionCtx)
+	go e.receiveFromServer(serverConn, e.clientConn, serverErrCh)
+
+	select {
+	case err := <-clientErrCh:
+		e.Log.WithError(err).Debug("Client done.")
+	case err := <-serverErrCh:
+		e.Log.WithError(err).Debug("Server done.")
+	case <-ctx.Done():
+		e.Log.Debug("Context canceled.")
 	}
 
 	return nil
 }
 
+// receiveFromClient relays protocol messages received from MySQL client
+// to MySQL database.
+func (e *Engine) receiveFromClient(clientConn, serverConn io.ReadWriteCloser, clientErrCh chan<- error, sessionCtx *common.Session) {
+	defer func() {
+		if r := recover(); r != nil {
+			e.Log.Warnf("Recovered while handling DB connection %v.", r)
+			err := trace.BadParameter("failed to handle client connection")
+			e.SendError(err)
+		}
+		serverConn.Close()
+		e.Log.Debug("Stop receiving from client.")
+		close(clientErrCh)
+	}()
+	for {
+		p, err := protocol.ReadPacket(clientConn)
+		if err != nil {
+			if utils.IsOKNetworkError(err) {
+				e.Log.Debug("Client connection closed.")
+				return
+			}
+			e.Log.WithError(err).Error("Failed to read client packet.")
+			clientErrCh <- err
+			return
+		}
+
+		packet, err := protocol.ToSQLPacket(p)
+		switch {
+		case err != nil:
+			e.Log.WithError(err).Errorf("Failed to parse SQLServer packet.")
```

Comment:
I wonder if we should emit some audit event in this case. In case someone sends a packet that we recognize (SQLBatch or RPCRequest) but malformed on purpose. Cause things like usually get highlighted in security audits as audit bypass.

Maybe intro a new event type, like DatabaseMalformedPacket or something, and use in cases like this?